Improvement/user react query in volume page

gaudiauj opened the pull request:
**Component**:

volumes pages

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 

In order to have a unified experience arround all the scality product, it was decided to use react-query rather than redux / redux saga. 

**Summary**:

This PR is the begining of using react-query in the UI of metalk8s. 
The first step is to migrate all the api call from redux-saga to react-query but still use redux as the state manager.

This PR aim to change the refresh call in the volume page. 

The next steps are : 

- add integration test. 
- migrate the other call to react-query
- get the data directly from react-query without using redux

**Acceptance criteria**: 

same comportement than before

gaudiauj committed:
improvement: migrate the refresh volume to reactQuery

gaudiauj committed:
improvement: migrating to react-query

migrate the fetchcurrentVolumeObject in the volume page from redux saga to react-query

gaudiauj committed:
improvement: migrate to react-query

migrate fetch persitent volumes from redux saga to reactquery

gaudiauj committed:
improvement: use default timeout fo refresh

gaudiauj committed:
improvement : migrate to react-query

migrate currentVolumeStats from react-saga to reactquery in the volume page

gaudiauj committed:
improvement: fix typo

gaudiauj committed:
improvement: remove fetchCurrentVolumeStatsAction as it is already done in useFetchCurrentVolumeStats

gaudiauj committed:
improvement: migrate refresh node to react-query on the volume page

gaudiauj committed:
lint: remove useless line break

bert-e commented:
# Hello gaudiauj,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/user-react-query-in-volume-page` into
`development/2.11`
* [w/123.0/improvement/user-react-query-in-volume-page](https://github.com/scality/metalk8s/compare/development/123.0...w/123.0/improvement/user-react-query-in-volume-page) will be merged into `development/123.0` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @alexandre-allard-scality

* @gdemonet

* @thomasdanan

* @chengyanjin

* @lucieleonard

* @teddyandrieux

* @jbwatenbergscality

* @nicolast

JBWatenbergScality commented on ui/src/containers/AlertProvider.js:
```diff
@@ -43,13 +44,18 @@ const InternalAlertProvider = ({
+
   return (
-    <FederatedComponent
+    <QueryClientProvider client={queryClient}>
```
We are using QueryClientProvider of shell ui in order to share cache with other UIs. I think the pb you met is likely due to the update not being performed on shell-ui side.

JBWatenbergScality commented on ui/src/ducks/app/volumes.hooks.js:
```diff
@@ -0,0 +1,28 @@
+      select: (data) => data.body?.items,
+      staleTime: fiveSecondsInMs,
+      refetchInterval: TenSecondsInMs,
+    },
```
I think we already have a query to retrieve the volumes in shell-ui platform library. This query should reuse the same  query to mutualise the cache.

JBWatenbergScality commented on ui/src/ducks/app/volumes.hooks.js:
```diff
@@ -0,0 +1,28 @@
+    },
+  );
+
+  useEffect(() => {
+    if (data) {
+      dispatch(setVolumesAction(data));
+    }
+  }, [data, dispatch]);
```
I think we can try to drop redux here and then I would question the added value of this custom hook. To me it would make sense to have a `useVolumes()` hook but then it's returned data should match the target format (including usage from prometheus and so on).

JBWatenbergScality commented on ui/src/ducks/app/volumes.hooks.js:
```diff
@@ -20,9 +20,34 @@ export function useRefreshVolume() {
+  return result;
+}
+
+export function useFetchCurrentVolumeObject(volumeName: string) {
```
I'm not sure refactoring using "simple" Fetch hooks is the right approach. I would more see what is the expected data output, what is the input and then build a hook abstracting away fetching, data conversion and formatting. Kind of what I tried to initialize with the `useVolumesWithAlerts` hook.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @alexandre-allard-scality

* @gdemonet

* @thomasdanan

* @chengyanjin

* @lucieleonard

* @teddyandrieux

* @jbwatenbergscality

* @nicolast





The following reviewers are expecting changes from the author, or must review again:

* @jbwatenbergscality

gaudiauj committed:
improvement: update react-query on both shell-ui and ui

gaudiauj committed:
improvement: create factory for query option

create factory for query option that would allow reuse of the option more easily in other files.

gaudiauj committed:
bugfix: fix missing spec data in persitent volume

ChengYanJin commented on ui/package-lock.json:
```diff
@@ -17885,9 +17885,9 @@
-      "version": "3.33.7",
-      "resolved": "https://registry.npmjs.org/react-query/-/react-query-3.33.7.tgz",
-      "integrity": "sha512-uL/1ZatxcRyvfhbNPeuOlLVbcdMQKzQ2t+SQLJMrMM0plfa/Al3buaPbGsbsmIwvyWeI4CVkkSBKzQIkU4p9MQ==",
+      "version": "3.34.0",
```
Likely there is a change in ui/package.json too?

Oh. I see it in the previous commits. 
Would ease the review process if we can have clean commits ;)

ChengYanJin replied:
I guess we miss the change in `ui/package.json` ;)

And since we bump the version of React Query in shell UI, we have to check if it's still compatible with other federated UIs. Because the other federated UI won't have the same version with shell after we merge your PR.

ChengYanJin commented on ui/src/ducks/app/monitoring.hooks.js:
```diff
@@ -1,48 +1,61 @@
-  const dispatch = useDispatch();
-
+export function getVolumeLatencyCurrentQueryOption() {
   const volumeLatencyCurrentQuery = `irate(node_disk_io_time_seconds_total[1h]) * 1000000`;
```
(Maybe not directly related to your PR)

I think the correct PromQL query would be `irate(node_disk_io_time_seconds_total[1h])`
The reason we add  ***1000000** is that we hardcode **ms** as the unit of the latency. 

It's a very bad practice IMO, the unit should be added based on the value.

ChengYanJin commented on ui/src/services/k8s/volumes.key.js:
```diff
@@ -0,0 +1,5 @@
+export const volumesKey = {
+  all: ['volumes', 'list'],
+  volumeObject: ['volumes', 'object'],
+  persitant: ['volumes', 'Persistant'],
```
persistentVolumes:  [''persistentVolumes"]

ChengYanJin commented on ui/src/ducks/app/monitoring.hooks.js:
```diff
@@ -1,48 +1,61 @@
-export function useFetchCurrentVolumeStats() {
-  const dispatch = useDispatch();
-
+export function getVolumeLatencyCurrentQueryOption() {
```
I don't think it makes sense to keep the functions that return the query option under the `/ducks/` folder.
  
We probably should put them under `/services/queries/volumes`
WDYT @JBWatenbergScality @MonPote ?

ChengYanJin commented on ui/src/services/k8s/core.key.js:
```diff
@@ -0,0 +1,3 @@
+export const nodeKey = {
```
The keys should live next to the queries too.

ChengYanJin commented on ui/src/ducks/app/nodes.hook.js:
```diff
@@ -0,0 +1,97 @@
+import { useEffect } from 'react';
```
I am not convinced to put the hook under `/src/ducks/`

ChengYanJin reviewed (dismissed):
Some comments about codebase structure.
We tend to put everything related to redux under `/ducks` folder, that's why I am not convinced to add queries and hooks here.

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/user-react-query-in-volume-page` has diverged from
`development/2.11` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.11` into `improvement/user-react-query-in-volume-page`
* Rebase `improvement/user-react-query-in-volume-page` onto `origin/development/2.11`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

gaudiauj committed:
improvement: move query and key to their own file in platformLibrary directory

bert-e commented:
# History mismatch



Merge commit #cb81ae43929b6063444968f65e52546759474986 on the integration branch
`w/123.0/improvement/user-react-query-in-volume-page` is merging a branch which is neither the current
branch `improvement/user-react-query-in-volume-page` nor the development branch
`development/123.0`.

It is likely due to a rebase of the branch `improvement/user-react-query-in-volume-page` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gaudiauj commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/user-react-query-in-volume-page` into
`development/2.11`
* [w/123.0/improvement/user-react-query-in-volume-page](https://github.com/scality/metalk8s/compare/development/123.0...w/123.0/improvement/user-react-query-in-volume-page) will be merged into `development/123.0` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @alexandre-allard-scality

* @gdemonet

* @thomasdanan

* @chengyanjin

* @lucieleonard

* @teddyandrieux

* @jbwatenbergscality

* @nicolast





The following reviewers are expecting changes from the author, or must review again:

* @chengyanjin

* @jbwatenbergscality

gaudiauj committed:
bugfix: ui fix missing param in react-query key

JBWatenbergScality commented on ui/src/containers/VolumePage.js:
```diff
@@ -31,27 +26,16 @@ const VolumePage = (props) => {
-    stopRefreshCurrentVolumeStatsAction,
-  );
+  useFetchCurrentVolumeObject(currentVolumeName);
+  useRefreshNodes();
```
Can we drop `dispatch(fetchNodesAction());` below ?
Can we `const nodes = useRefreshNodes();` and drop `const nodes = useTypedSelector((state) => state.app.nodes.list);` below ?
I wonder the same for the other resources.

gaudiauj replied:
we have actually an other tickets to handle the removal of all redux function

JBWatenbergScality commented on ui/src/hooks/nodes.js:
```diff
@@ -0,0 +1,106 @@
+      dispatch(updateNodesAction({ list: data }));
+
+      // To make sure that the loader is visible for at least 1s
+      setTimeout(function () {
+        // this dispatch should be migrated to react-query
+        // however it should be done when the nodes pages is migrated
+        dispatch({ type: FETCH_NODES_IPS_INTERFACES });
+      }, 1000);
```
This timeout should be cleaned up to avoid memory leaks.

gaudiauj committed:
bugfix: clear setTimeout

JBWatenbergScality commented on ui/src/hooks/nodes.js:
```diff
@@ -0,0 +1,106 @@
+  const { data, isLoading } = result;
+
+  useEffect(() => {
+    function dispatchIpInterfaces() {
+      dispatch({ type: FETCH_NODES_IPS_INTERFACES });
+    }
+    if (data) {
+      dispatch(updateNodesAction({ list: data }));
+
+      // To make sure that the loader is visible for at least 1s
+      setTimeout(dispatchIpInterfaces, 1000);
+    }
+    dispatch(updateNodesAction({ isLoading }));
+    return () => clearTimeout(dispatchIpInterfaces);
```
```suggestion
    let timer = false;
    if (data) {
      dispatch(updateNodesAction({ list: data }));

      // To make sure that the loader is visible for at least 1s
      const timer = setTimeout(() => {
          dispatch({ type: FETCH_NODES_IPS_INTERFACES });
      }, 1000);
    }
    dispatch(updateNodesAction({ isLoading }));
    return () => {
        if (timer) {
            clearTimeout(timer);
        }
     }
```

JBWatenbergScality replied:
https://developer.mozilla.org/en-US/docs/Web/API/clearTimeout

bert-e commented:
# History mismatch



Merge commit #b0f675a536d74a18c460282b0eef5a4e667ee547 on the integration branch
`w/123.0/improvement/user-react-query-in-volume-page` is merging a branch which is neither the current
branch `improvement/user-react-query-in-volume-page` nor the development branch
`development/123.0`.

It is likely due to a rebase of the branch `improvement/user-react-query-in-volume-page` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

JBWatenbergScality commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/user-react-query-in-volume-page` into
`development/2.11`
* [w/123.0/improvement/user-react-query-in-volume-page](https://github.com/scality/metalk8s/compare/development/123.0...w/123.0/improvement/user-react-query-in-volume-page) will be merged into `development/123.0` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @gdemonet

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality





The following reviewers are expecting changes from the author, or must review again:

* @chengyanjin

gaudiauj commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @gdemonet

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality





The following reviewers are expecting changes from the author, or must review again:

* @chengyanjin







*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/37d8a65d1e5785456f7de28d52ed39681b65c9b7) did not succeed in branch w/123.0/improvement/user-react-query-in-volume-page.




*The following options are set:* **approve**

gaudiauj commented:
/approve

gaudiauj committed:
bugfix: fix potential infinite rerender loop in volume page

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @gdemonet

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality










*The following options are set:* **approve**

gaudiauj commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/user-react-query-in-volume-page` into
`development/2.11`
* [w/123.0/improvement/user-react-query-in-volume-page](https://github.com/scality/metalk8s/compare/development/123.0...w/123.0/improvement/user-react-query-in-volume-page) will be merged into `development/123.0` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @gdemonet

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality










*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/6be2ffbcdd6d50177e5763aeaff160609c490bbd) did not succeed in branch w/123.0/improvement/user-react-query-in-volume-page.




*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #867b184672fc066e917775cbd353737cc40068bd on the integration branch
`w/123.0/improvement/user-react-query-in-volume-page` is merging a branch which is neither the current
branch `improvement/user-react-query-in-volume-page` nor the development branch
`development/123.0`.

It is likely due to a rebase of the branch `improvement/user-react-query-in-volume-page` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

gaudiauj commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/user-react-query-in-volume-page` into
`development/2.11`
* [w/123.0/improvement/user-react-query-in-volume-page](https://github.com/scality/metalk8s/compare/development/123.0...w/123.0/improvement/user-react-query-in-volume-page) will be merged into `development/123.0` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @chengyanjin

* @lucieleonard

* @jbwatenbergscality

* @alexandre-allard-scality

* @gdemonet

* @teddyandrieux

* @nicolast

* @thomasdanan










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.11`

* :heavy_check_mark: `development/123.0`



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.11`

* :heavy_check_mark: `development/123.0`



The following branches have **NOT** changed:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



Please check the status of the associated issue None.

Goodbye gaudiauj.
